tests/ModelContextProtocol.Tests/Server/McpServerToolTests.cs

```diff
@@ -45,20 +45,47 @@ public async Task SupportsIMcpServer()
         Assert.Equal("42", result.Content[0].Text);
     }
 
-    [Fact]
-    public async Task SupportsServiceFromDI()
+    [Theory]
+    [InlineData(ServiceLifetime.Singleton)]
+    [InlineData(ServiceLifetime.Scoped)]
+    [InlineData(ServiceLifetime.Transient)]
+    public async Task SupportsServiceFromDI(ServiceLifetime injectedArgumentLifetime)
     {
-        MyService expectedMyService = new();
+        MyService singletonService = new();
 
         ServiceCollection sc = new();
-        sc.AddSingleton(expectedMyService);
-        IServiceProvider services = sc.BuildServiceProvider();
+        switch (injectedArgumentLifetime)
+        {
+            case ServiceLifetime.Singleton:
+                sc.AddSingleton(singletonService);
+                break;
+
+            case ServiceLifetime.Scoped:
+                sc.AddScoped(_ => new MyService());
+                break;
+
+            case ServiceLifetime.Transient:
+                sc.AddTransient(_ => new MyService());
+                break;
+        }
 
-        McpServerTool tool = McpServerTool.Create((MyService actualMyService) =>
+        sc.AddSingleton(services =>
         {
-            Assert.Same(expectedMyService, actualMyService);
-            return "42";
-        }, new() { Services = services });
+            return McpServerTool.Create((MyService actualMyService) =>
+            {
+                Assert.NotNull(actualMyService);
+                if (injectedArgumentLifetime == ServiceLifetime.Singleton)
+                {
+                    Assert.Same(singletonService, actualMyService);
+                }
+
+                return "42";
+            }, new() { Services = services });
+        });
+
+        IServiceProvider services = sc.BuildServiceProvider();
+
+        McpServerTool tool = services.GetRequiredService<McpServerTool>();
```

Comment:
```suggestion
        IServiceProvider services = sc.BuildServiceProvider(validateScopes: true);

        await using var scope = services.CreateAsyncScope();
        McpServerTool tool = scope.ServiceProvider.GetRequiredService<McpServerTool>();
```

Nit: Not a big deal since this is just a test, but resolving a scoped service without first creating a scope is bad practice, and our host's now guard against it in dev environments. https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/9.0/hostbuilder-validation

Again, it doesn't really matter, but it's good to always show resolving scoped services from a scope. We might need to update `SetupGet(x => x.Services).Returns(scope.ServiceProvider);` below.

Reply:
> resolving a scoped service without first creating a scope is bad practice

Is there a product change we need here, to ensure the IServiceProvider passed to the tool's invocation is appropriately scoped? I'm looking at https://github.com/modelcontextprotocol/csharp-sdk/issues/198 wondering how to repro that, but maybe we need to be creating a scope as part of a stdio server's handling?

Reply:
> unless you disable validation

Where is that validation coming from / why is my test above not hitting it?

Reply:
The validation is coming from the host in "development" environments from [here](https://github.com/dotnet/runtime/blob/778427c4c7618a753f4d2afb75943590c3f2d8f3/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L330). This used to always default to false until .NET 9.

I also left a comment on #198 after making my last comment here. After further though, I do think we should make product changes. We should create a scope per handler invocation. I assigned the issue to myself.

Reply:
> The validation is coming from the host in "development" environments from [here](https://github.com/dotnet/runtime/blob/778427c4c7618a753f4d2afb75943590c3f2d8f3/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L330). This used to always default to false until .NET 9.

Ok, so this test isn't hitting that because it's not using a host.